src/rtypeconverter.cc

```diff
@@ -300,3 +299,222 @@ void ConvertToProtoBuf::compositeToProtoBuf(SEXP v, std::vector<PlcDataType> &su
+                                            type);
+                            break;
+                    }
+                } else {
+                    element->set_isnull(true);
+                }
+            }
+        } break;
+        case REALSXP: {
+            for (int i = 0; i < length; i++) {
+                plcontainer::ScalarData *element = result->add_values();
+                if (REAL(v)[i] != NA_REAL) {
+                    switch (type) {
+                        case plcontainer::PlcDataType::LOGICAL: {
+                            element->set_logicalvalue((bool)REAL(v)[i] == 0 ? 0 : 1);
+                        } break;
+                        case plcontainer::PlcDataType::INT: {
+                            element->set_intvalue((int32_t)REAL(v)[i]);
+                        } break;
+                        case plcontainer::PlcDataType::REAL: {
+                            element->set_realvalue(REAL(v)[i]);
+                        } break;
+                        case plcontainer::PlcDataType::TEXT: {
+                            element->set_stringvalue(std::to_string(REAL(v)[i]));
+                        } break;
+                        case plcontainer::PlcDataType::BYTEA: {
+                            element->set_byteavalue(std::to_string(REAL(v)[i]));
+                        } break;
+                        default:
+                            this->rLog->log(RServerLogLevel::WARNINGS, "Unsupport type in value %d",
+                                            type);
+                            break;
+                    }
+                } else {
+                    element->set_isnull(true);
+                }
+            }
+        } break;
+        case STRSXP: {
+            for (int i = 0; i < length; i++) {
+                const char *value = CHAR(STRING_ELT(v, i));
+                plcontainer::ScalarData *element = result->add_values();
+                if (STRING_ELT(v, i) != NA_STRING && value != NULL) {
+                    switch (type) {
+                        case plcontainer::PlcDataType::TEXT: {
+                            element->set_stringvalue(std::string(value));
+                        } break;
+                        case plcontainer::PlcDataType::BYTEA: {
+                            element->set_byteavalue(std::string(value));
+                        } break;
+                        default:
+                            this->rLog->log(RServerLogLevel::WARNINGS, "Unsupport type in value %d",
+                                            type);
+                            break;
+                    }
+                } else {
+                    element->set_isnull(true);
+                }
+            }
+        } break;
+        default:
+            this->rLog->log(RServerLogLevel::WARNINGS, "Unsupport type in value %d", type);
+            break;
+    }
+}
+
+SEXP ConvertToSEXP::allocRVector(const plcontainer::PlcDataType &type, int length) {
```

Comment:
PlcDataType is enum, so it's better to pass by value than const reference. There are several similar issues in other functions.

Reply:
Thx, fixed.